rust/protocol/src/error.rs

```diff
@@ -11,82 +15,208 @@ use uuid::Uuid;
+    /// signatures to minimize the need to raise this error to FFI boundaries.
     InvalidArgument(String),
     /// invalid state for call to {0} to succeed: {1}
+    ///
+    /// Raised if some optional value was missing before performing some operation which needed it.
+    ///
+    /// Prefer to avoid returning [std::result::Result] and [Option] from struct methods in cases
+    /// where they're not necessary for trait polymorphism, as well as using dedicated wrapper
+    /// structs in API signatures, to minimize the need to raise this error.
     InvalidState(&'static str, String),
 
     /// protobuf encoding was invalid
+    ///
+    /// Raised if a field in a protobuf is invalid in some way.
+    ///
+    /// Prefer to raise [Self::InvalidState] except in methods which directly decode protobufs.
     InvalidProtobufEncoding,
 
     /// ciphertext serialized bytes were too short <{0}>
+    ///
+    /// Raised if some ciphertext was deserialized from a too-small slice.
+    ///
+    /// Prefer to make API method signatures and wrapper structs consume and produce static-sized
+    /// byte slices to minimize the need to raise this error.
     CiphertextMessageTooShort(usize),
     /// ciphertext version was too old <{0}>
+    ///
+    /// Raised if the ciphertext version decoded from a protobuf is older than this client.
+    ///
+    /// The current client's ciphertext version is at [CIPHERTEXT_MESSAGE_CURRENT_VERSION].
     LegacyCiphertextVersion(u8),
     /// ciphertext version was unrecognized <{0}>
+    ///
+    /// Raised if the ciphertext version decoded from a protobuf is newer than this client.
+    ///
+    /// The current client's ciphertext version is at [CIPHERTEXT_MESSAGE_CURRENT_VERSION].
     UnrecognizedCiphertextVersion(u8),
     /// unrecognized message version <{0}>
+    ///
+    /// Raised if the ciphertext version decoded from a protobuf fails to match the cached version
+    /// for the message chain that message originates from.
+    ///
+    /// *TODO: This case should wrap the same numeric type as [Self::LegacyCiphertextVersion] and
+    /// [Self::UnrecognizedCiphertextVersion]. This dissonance is addressed in
+    /// <https://github.com/signalapp/libsignal-client/pull/289>.*
     UnrecognizedMessageVersion(u32),
 
     /// fingerprint version number mismatch them {0} us {1}
+    ///
+    /// Raised if a fingerprint version decoded from a protobuf has an unexpected value.
     FingerprintVersionMismatch(u32, u32),
     /// fingerprint parsing error
+    ///
+    /// Raised if a field in a fingerprint protobuf is invalid in some way.
+    ///
+    /// Similar to [Self::InvalidProtobufEncoding].
     FingerprintParsingError,
 
     /// no key type identifier
+    ///
+    /// Raised if a [PublicKey] is deserialized from an empty slice.
+    ///
+    /// Prefer to use static-sized slices in API method signatures and struct fields to minimize the
+    /// need to raise this error.
     NoKeyTypeIdentifier,
```

Comment:
Yeah, this is a little silly. No one should mistake an empty slice for a public key.